aptos-move/framework/aptos-framework/sources/scheduled_transactions/scheduled_txns.move

```diff
@@ -0,0 +1,1208 @@
+        );
+        cancel_internal(signer::address_of(sender), key, deposit_amt);
+    }
+
+    const MASK_64: u256 = 0xffffffffffffffff; // 2^64 - 1
+
+    fun truncate_to_u64(val: u256): u64 {
+        let masked = val & MASK_64; // Truncate high bits
+        (masked as u64) // Now safe: always <= u64::MAX
+    }
+
+    // Converting 32-byte hash to u256
+    fun hash_to_u256(hash: vector<u8>): u256 {
+        assert!(hash.length() == 32, error::internal(EINVALID_HASH_SIZE));
+        from_bcs::to_u256(hash)
+    }
+
+    /// Internal cancel function that takes an address instead of signer. No signer verification, assumes key is present
+    /// in the schedule_map.
+    fun cancel_internal(
+        account_addr: address, key: ScheduleMapKey, deposit_amt: u64
+    ) acquires ScheduleQueue, AuxiliaryData {
+        let queue = borrow_global_mut<ScheduleQueue>(@aptos_framework);
+
+        // Remove the transaction from schedule_map and txn_table
+        queue.schedule_map.remove(&key);
+        queue.txn_table.remove(key.txn_id);
+
+        // Refund the deposit
+        // Get owner signer from capability
+        let gas_deposit_store_cap = borrow_global<AuxiliaryData>(@aptos_framework);
+        let gas_deposit_store_signer =
+            account::create_signer_with_capability(
+                &gas_deposit_store_cap.gas_fee_deposit_store_signer_cap
+            );
+
+        // Refund deposit from owner's store to sender
+        coin::transfer<AptosCoin>(
+            &gas_deposit_store_signer,
+            account_addr,
+            deposit_amt
+        );
+    }
+
+    /// Gets txns due to be run; also expire txns that could not be run for a while (mostly due to low gas priority)
+    fun get_ready_transactions(
+        block_timestamp_ms: u64
+    ): vector<ScheduledTransactionInfoWithKey> acquires ScheduleQueue, AuxiliaryData, ToRemoveTbl {
+        get_ready_transactions_with_limit(
+            block_timestamp_ms, GET_READY_TRANSACTIONS_LIMIT
+        )
+    }
+
+    fun get_ready_transactions_with_limit(
+        block_timestamp_ms: u64, limit: u64
+    ): vector<ScheduledTransactionInfoWithKey> acquires ScheduleQueue, AuxiliaryData, ToRemoveTbl {
+        remove_txns(block_timestamp_ms);
+        // If scheduling is shutdown, we cannot schedule any more transactions
+        let aux_data = borrow_global<AuxiliaryData>(@aptos_framework);
+        if (aux_data.module_status != ScheduledTxnsModuleStatus::Active) {
+            return vector::empty<ScheduledTransactionInfoWithKey>();
+        };
+
+        let queue = borrow_global<ScheduleQueue>(@aptos_framework);
+        let scheduled_txns = vector::empty<ScheduledTransactionInfoWithKey>();
+        let count = 0;
+
+        let iter = queue.schedule_map.new_begin_iter();
+        while ((count < limit) && !iter.iter_is_end(&queue.schedule_map)) {
+            let key = iter.iter_borrow_key();
+            if (key.time > block_timestamp_ms) {
```

Comment:
for the `key.time == block_timestamp_ms` case, it's possible that we schedule a txn earlier than it's desired time. (because we convert the block_timestamp from us to ms).

I'd suggest we either change this to `>=`, or we use microseconds for calculation.

Reply:
Let's say block timestamp is 1000001 (in us), scheduled time is 1000 (in ms), after converting to ms it would be 1000 for both so the txn will be run. (even when block timestamp is greater than the scheduled time in us)